daemon/create_unix.go

```diff
@@ -75,6 +81,22 @@ func (daemon *Daemon) createContainerOSSpecificSettings(container *container.Con
 	return daemon.populateVolumes(container)
 }
 
+func canDirectSetBaseFS(container *container.Container, config *containertypes.Config, hostConfig *containertypes.HostConfig) bool {
+	if container.Driver != "overlay" && container.Driver != "overlay2" {
+		return false
+	}
+	if len(config.Volumes) != 0 {
+		return false
+	}
+	if len(hostConfig.Binds) != 0 {
+		return false
+	}
+	if config.WorkingDir != "" {
+		return false
+	}
+	return true
+}
```

Comment:
This seems _really_ limited -- is the speedup really enough better to warrant the additional code complexity?

Reply:
Yes, It's not available for many  casese. I set it in WIP now.

Reply:
Yes, it concerns me that

- the daemon has to use storage-driver dependent logic (so it needs to be aware of specific storage-driver's behaviour, which defeats the idea of a "driver" with an interface 🤔)
- the list of conditions above seems to be "these are conditions we are _aware of_ that they won't work". That seems brittle; new conditions that don't work may arrive. e.g., from the above, it looks like it can't be done if the container has mounts/binds specified; does this also apply for implicit mounts? (e.g. `/etc/resolv.conf` mounted from the host, etc?)
- having a quick look at `daemon.Mount()`, I see it's also applying a mount-label to the mount; can that safely be skipped? https://github.com/moby/moby/blob/b88acf7a7a571b0fe054b8315c79caa16b0c92df/daemon/daemon.go#L1282


I'm curious as well how significant the performance gain is; do you have have benchmarks?

Reply:
Alternatively you could add a function to the graph driver to get the path.